### DISCUSSION TITLE
New graph type for position tracing - How To?

### DISCUSSION THREAD

**Discussion by DRNadler** [Blake-Madden/Wisteria-Dataviz#47] [Q&A]
I'm considering using Wysteria for a time-series analysis tool I must write (detailed requirements for complete tool below).
_**Let me know if this is a bad idea!**_

Any suggestions on how to implement a track visualization graph in Wisteria? I will display recorded 3D vehicle track data 3D (coordinates and altitude). For each point also tiny vertical graph of another parameter p. I can prescale points to nominal display xyz plus p. But where to start implementing such a graph type? 

Thanks in advance,
Best Regards, Dave

### **Analysis Tool Requirements**

**Time-series analysis**
-    load a file containing multiple time-series (GPS track plus sensors)
-    compute derived data time-series (algorithm development and testing)
-    visualize interesting time-ranges
-    mark and annotate interesting time-ranges
-    save setup and notes on interesting display ranges for later replay

**Visualization**

**User controls for time-series**
-    Which time series to display (from the set of recorded and derived values).
-    Time range to display, ideally via a control window with time zoom and range select (viewport control)

**Time-series display**
-    selected time-series, with common time-scale (X-axis) and separate Y-axis labeling
-    tool-tip detail (mouse on graph shows time series name and numeric value)
-    time-select bar (select a time-range and a specific point in the time-range for separate detail visualization, ie highlight point in GPS vehicle track display)

**GPS vehicle track visualization**
- GPS track trace
-  window scaled to contain range of coordinates in selected time-range
- 3D option with perspective
- tiny vertical bar  (positive green some pixels up, negative red some pixels down)

**Annotation and Recording/Replay**
To support ongoing analysis of interesting display regions, annotate interesting display regions and save the set of annotations. A saved session can be later loaded and further annotated and saved...
An analytic session contains (during runtime, or save and restore to/from a JSON file):
-    time-series file name
-    list of interesting ranges, each containing:
--        list of time-series to display
--        time range
--        current selected time (within time range)
--        annotation (textual comment)

**Comment by Blake-Madden** [comment]
I like the idea of *some day* having 3D support, but to be honest it would be a very long time before I could explore something like that. Plus, it would be rather simple, such as viewing it from a fixed 45 degree perspective. Being able to change the perspective (and rotate it) probably wouldn't happen.

**Comment by DRNadler** [comment]
OK, I might not do variable perspective. Where do I start to create an XY plot (vehicle may move in a circle etc)?
IE more than one Y value per X...

**Reply by Blake-Madden** [reply]
You can add custom Y axes, that could give you multiple Y axes for one X.

If you could upload a mock up (a hand drawing would be fine) to visualize what you are looking for, that might help. Sorry, I'm having trouble visualizing what the final graphic would be here.

**Comment by DRNadler** [comment]
Below is a first try using wxMathPlot. I need to add a tiny vertical plot, for example at each 10th point, showing a non-track data parameter. Also this plot is only 2D, no 3rd dimension yet, I need a viewpoint projection.

Interaction with the plot:
- mouse-wheel zoom in and out (working),
- right-click about point:
-- other information about the point (additional parameters dialog or somesuch)
-- adjust plot start-stop range (viewport) to this point
-- maybe adjust viewpoint and projection

Hope I'm making some sense...

![image](https://github.com/user-attachments/assets/51f8b7f9-5e4f-42ee-b309-642ceca45e68)

**Comment by Blake-Madden** [comment]
I *think* the 2D line plot will work for you, because it plots the points along the X axis in the order that you provide them. In other words, I do **not** sort the values by their X positions, I draw a line along the data points in the order that you provide them.

A side effect (feature?) of this is that a line can go back and forth, and even in circles. Please refer to the W-Curve plot (which is a simple derivation of 2D line plot):

https://github.com/Blake-Madden/Wisteria-Dataviz/blob/main/docs/images/WCurve.svg

In this image, note how the line for "Frank" goes backwards and spirals around. This is achieved from the actual data and requesting that the line be drawn as a spline (just to make it smoother).

Does that help? Is that sort of what you are needing?

**Comment by DRNadler** [comment]
Looks like that may work, sorry I didn't notice that in the documentation... 
I'll try to test next week, Thanks!
